lib/web/apiserver.go

```diff
@@ -5162,15 +5164,105 @@ func (h *Handler) authExportPublic(w http.ResponseWriter, r *http.Request, p htt
 	)
 	if err != nil {
 		h.log.WithError(err).Debug("Failed to generate CA Certs.")
-		http.Error(w, err.Error(), trace.ErrorToCode(err))
+		h.serveErrorResponse(err, w, r)
 		return
 	}
 
-	reader := strings.NewReader(authorities)
+	err = serveTextResponse(authorities, w, r)
+	if err != nil {
+		h.log.WithError(err).Error("Failed to encode and send response.")
+		h.serveErrorResponse(err, w, r)
+	}
+}
+
+// Given an HTTP request and a response body to the request, return an encoded response based
+// on the `Accept` header, as well as the appropriate Content-Type.
+// If an error has occured, the response encoding and content type are chosen on a best-effort basis.
+// Will always return encoded response bytes, and content type value.
+func encodeTextResponse(responseBody string, r *http.Request) ([]byte, string, error) {
+	// Track the latest error. Callers can choose to handle or ignore this. Useful for
+	// error handling functions.
+	var trackedErr error
+
+	// Parse "Accept" header values into mime types
+	var mimeTypes []string
+	// Handle the case of multiple headers
+	for _, acceptHeaderValue := range r.Header["Accept"] {
+		// Handle the case of multiple values in a single header
+		for _, splitHeaderValue := range strings.Split(acceptHeaderValue, ",") {
+			splitHeaderValue = strings.TrimSpace(splitHeaderValue)
+
+			// Parameters (such as charset and quality factor) are not currently supported
+			var mimeType string
+			mimeType, _, err := mime.ParseMediaType(splitHeaderValue)
+			if err != nil {
+				trackedErr = trace.Wrap(err, "malformed mime type %q", splitHeaderValue)
+				continue
+			}
+
+			mimeTypes = append(mimeTypes, mimeType)
+		}
+	}
+
+	// Ignore duplicate MIME types
+	mimeTypes = slices.Compact(mimeTypes)
+
+	// Add a default value in the event that either no value is sent, or no supported value is sent
+	// This must always be a plain text response to preserve backwards compatibility
+	mimeTypes = append(mimeTypes, "text/plain")
 
-	// ServeContent sets the correct headers: Content-Type, Content-Length and Accept-Ranges.
+	for _, mimeType := range mimeTypes {
+		switch mimeType {
+		case "application/json":
+			var responseContent []byte
+			responseContent, err := json.Marshal(responseBody)
+			if err != nil {
+				trackedErr = trace.Wrap(err, "failed to marshal JSON response")
+				continue
+			}
+
+			return responseContent, "application/json", trackedErr
+		// Default to text/plain
+		case "text/plain":
+			return []byte(responseBody), "text/plain; charset=utf-8", trackedErr
+		}
+	}
+
+	// This will never be hit but the Golang compiler isn't smart enough to realize this will never be hit
+	return nil, "", trace.Errorf("failed to match MIME type (Teleport bug - this should never be hit)")
```

Comment:
The error message is not very helpful for the end user (which would be fine), but, as I said above, you can move the `text/plain` branch and use it as the default condition.

Reply:
Unfortunately this won't work - see conversation above. And even with a `default` statement and one entry in the array, the compiler isn't smart enough to know that the error line will never be hit.